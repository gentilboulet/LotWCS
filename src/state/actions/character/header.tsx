import { createAction } from "typesafe-actions";

import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";
import { IPerk } from "../../../perks";

export const setName = createAction("header/SET_NAME", (name: string) => {
  return { name };
})();

export const setConcept = createAction(
  "header/SET_CONCEPT",
  (concept: string) => {
    return { concept };
  },
)();

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return { archetype };
  },
)();

export const setRank = createAction(
  "header/SET_RANK",
  (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    const dr = dataRanks.getRank(rank) as dataRanks.IDataRank;
    return { rank, perks: dr.perks as IPerk[] };
  },
)();
