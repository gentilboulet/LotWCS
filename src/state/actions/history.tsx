import * as constants from 'state/constants/history';

export interface IhistoryDeleteUpTo {
  type: constants.HISTORY_DELETE;
  id: number;
}

export type IHistoryAction = IhistoryDeleteUpTo;

export function historyDeleteUpTo(id: number): IhistoryDeleteUpTo {
  return { type: constants.HISTORY_DELETE, id };
}
