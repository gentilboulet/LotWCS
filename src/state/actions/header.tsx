import * as constants from 'state/constants/header';

import * as dataArchetypes from 'data/archetypes';
import * as dataRanks from 'data/ranks';

export interface IHeaderSetName {
  type: constants.HEADER_SET_NAME;
  name: string;
}

export interface IHeaderSetConcept {
  type: constants.HEADER_SET_CONCEPT;
  concept: string;
}

export interface IHeaderSetArchetype {
  type: constants.HEADER_SET_ARCHETYPE;
  archetype: string;
}

export interface IHeaderSetRank {
  type: constants.HEADER_SET_RANK;
  rank: string;
}

export type IHeaderAction = IHeaderSetName | IHeaderSetConcept | IHeaderSetArchetype | IHeaderSetRank;

export function setName(name: string): IHeaderSetName {
  return { type: constants.HEADER_SET_NAME, name };
}

export function setConcept(concept: string): IHeaderSetConcept {
  return { type: constants.HEADER_SET_CONCEPT, concept };
}

export function setArchetype(archetype: string): IHeaderSetArchetype {
  dataArchetypes.validateArchetype(archetype);
  return { archetype, type: constants.HEADER_SET_ARCHETYPE };
}

export function setRank(rank: string): IHeaderSetRank {
  dataRanks.validateRank(rank);
  return { rank, type: constants.HEADER_SET_RANK };
}
