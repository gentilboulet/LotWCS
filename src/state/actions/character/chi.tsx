import { createAction } from "typesafe-actions";
import { historyMetaCreator } from "../meta";
import { ICost } from "../../models/character/costs";
import { TChiName } from "../../../data/chi";

export const chiBuy = createAction(
  "chi/BUY",
  (chi: TChiName, value: number, cost: ICost) => {
    return { chi, value, cost };
  },
  historyMetaCreator,
)();
