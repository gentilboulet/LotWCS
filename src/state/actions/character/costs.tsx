import { createAction } from "typesafe-actions";
import { ICost } from "../../models/character/costs";

export const payCost = createAction("character/COST_PAY", (cost: ICost) => {
  return { cost };
})();
