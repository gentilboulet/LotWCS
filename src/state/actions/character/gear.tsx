import { createAction } from "typesafe-actions";

import { ARMOR_TYPE } from "../../../data/armor";
import { WEAPON_TYPE } from "../../../data/weapons";
import { ICost } from "../../models/character/costs";
import { historyMetaCreator } from "../meta";

export const armorBuy = createAction(
  "gear/ARMOR_BUY",
  (name: string, description: string, type: ARMOR_TYPE, cost: ICost) => {
    return { cost, description, name, type };
  },
  historyMetaCreator,
)();

export const armorCustomize = createAction(
  "gear/ARMOR_CUSTOM",
  (armorIdx: number, description: string) => {
    return { description, armorIdx };
  },
  historyMetaCreator,
)();

export const weaponBuy = createAction(
  "gear/WEAPON_BUY",
  (
    name: string,
    description: string,
    type: [WEAPON_TYPE] | [WEAPON_TYPE, WEAPON_TYPE],
    cost: ICost,
  ) => {
    return { cost, description, name, type };
  },
  historyMetaCreator,
)();

export const weaponCustomize = createAction(
  "gear/WEAPON_CUSTOM",
  (weaponIdx: number, description: string) => {
    return { description, weaponIdx };
  },
  historyMetaCreator,
)();
