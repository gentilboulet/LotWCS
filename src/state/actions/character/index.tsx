import { ActionType, isActionOf } from "typesafe-actions";

import * as chi from "./chi";
import * as cost from "./costs";
import * as gear from "./gear";
import * as header from "./header";
import * as kungfu from "./kungfu";
import * as loresheets from "./loresheets";
import * as perks from "./perks";
import * as skills from "./skills";
import * as virtues from "./virtues";

export type ICharacterAction =
  | ActionType<typeof chi>
  | ActionType<typeof cost>
  | ActionType<typeof gear>
  | ActionType<typeof header>
  | ActionType<typeof kungfu>
  | ActionType<typeof loresheets>
  | ActionType<typeof perks>
  | ActionType<typeof skills>
  | ActionType<typeof virtues>;

export const actions = {
  ...chi,
  ...cost,
  ...gear,
  ...header,
  ...kungfu,
  ...loresheets,
  ...perks,
  ...skills,
  ...virtues,
};

export const isCharacterAction = (
  action: any,
): ICharacterAction | undefined => {
  if (isActionOf(Object.values(actions), action)) {
    return action;
  }
  return undefined;
};
