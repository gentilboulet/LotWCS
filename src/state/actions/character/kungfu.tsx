import { createAction } from "typesafe-actions";

import * as dataKungFu from "../../../data/kungfu";
import { KUNGFU_TYPE } from "../../../data/kungfu/types";
import { bonusCultivation } from "../../../perks/actions/bonuses";
import { ICost } from "../../models/character/costs";
import { historyMetaCreator } from "../meta";

export const openStyle = createAction(
  "kungfu/OPEN_STYLE",
  (uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);

    return {
      cost,
      uid,
      kungfuType,
      perks: [
        bonusCultivation(
          dataKungFu.getKungfuChi(kungfuType, uid),
          cost.original,
        ),
      ],
    };
  },
  historyMetaCreator,
)();

export const buyTechnique = createAction(
  "kungfu/BUY_TECHNIQUE",
  (styleUid: string, uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    return {
      cost,
      kungfuType,
      styleUid,
      uid,
      perks: [
        bonusCultivation(
          dataKungFu.getKungfuChi(kungfuType, styleUid),
          cost.original,
        ),
      ],
    };
  },
  historyMetaCreator,
)();

export const customStyleName = createAction(
  "kungfu/CUSTOM_NAME_FOR_STYLE",
  (uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return {
      kungfuType,
      name,
      uid,
    };
  },
  historyMetaCreator,
)();

export const customTechniqueName = createAction(
  "kungfu/CUSTOM_NAME_FOR_TECHNIQUE",
  (styleUid: string, uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return {
      kungfuType,
      name,
      styleUid,
      uid,
    };
  },
  historyMetaCreator,
)();
