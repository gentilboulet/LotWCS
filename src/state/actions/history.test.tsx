import * as actions from "./history";

import { setName } from "../actions/character/header";

describe("Testing history action creators", () => {
  it("should create an history push", () => {
    const toPush = setName("bob");
    const a = actions.historyPush(toPush);
    expect(a.payload.action).toBe(toPush);
    expect(a).toMatchSnapshot();
  });
});
