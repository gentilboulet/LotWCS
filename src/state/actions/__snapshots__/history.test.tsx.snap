// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing history action creators should create an history delete action 1`] = `
Object {
  "id": 12,
  "type": "HISTORY_DELETE",
}
`;
